galley/pkg/config/analysis/analyzers/injection/image-auto.go

```diff
@@ -0,0 +1,131 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package injection
+
+import (
+	"strings"
+
+	admit_v1 "k8s.io/api/admissionregistration/v1"
+	apps_v1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	klabels "k8s.io/apimachinery/pkg/labels"
+
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/pkg/config/resource"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+)
+
+// ImageAutoAnalyzer reports an error if Pods and Deployments with `image: auto` are not going to be injected.
+type ImageAutoAnalyzer struct{}
+
+var _ analysis.Analyzer = &ImageAutoAnalyzer{}
+
+const manualInjectionImage = "auto"
+
+// Metadata implements Analyzer.
+func (a *ImageAutoAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name:        "injection.ImageAutoAnalyzer",
+		Description: "Makes sure that Pods and Deployments with `image: auto` are going to be injected",
+		Inputs: collection.Names{
+			collections.K8SCoreV1Namespaces.Name(),
+			collections.K8SCoreV1Pods.Name(),
+			collections.K8SAppsV1Deployments.Name(),
+			collections.K8SAdmissionregistrationK8SIoV1Mutatingwebhookconfigurations.Name(),
+		},
+	}
+}
+
+// Analyze implements Analyzer.
+func (a *ImageAutoAnalyzer) Analyze(c analysis.Context) {
+	var istioWebhooks []admit_v1.MutatingWebhook
+	c.ForEach(collections.K8SAdmissionregistrationK8SIoV1Mutatingwebhookconfigurations.Name(), func(resource *resource.Instance) bool {
+		mwhc := resource.Message.(*admit_v1.MutatingWebhookConfiguration)
+		for _, wh := range mwhc.Webhooks {
+			if strings.HasSuffix(wh.Name, "istio.io") {
+				istioWebhooks = append(istioWebhooks, wh)
+			}
+		}
+		return true
+	})
+	c.ForEach(collections.K8SCoreV1Pods.Name(), func(resource *resource.Instance) bool {
+		p := resource.Message.(*v1.Pod)
+		if !hasAutoImage(&p.Spec) {
+			return true
+		}
+		nsLabels := getNamespaceLabels(c, p.Namespace)
+		if !matchesWebhooks(nsLabels, p.Labels, istioWebhooks) {
+			m := msg.NewImageAutoWithoutInjectionError(resource, "Pod", p.Name)
+			c.Report(collections.K8SCoreV1Pods.Name(), m)
```

Comment:
Should always report if `hasAutoImage`, it doesn't matter if it matches the webhooks. If you have `image: auto` your pod is broken and will never repair itself - even if a webhook was added latter.

Reply:
Makes sense, changed